File: apk/app/src/main/java/com/example/momentra/data/create/IdempotencyKeyStore.kt
package com.example.momentra.data.create

import android.content.Context
import java.util.UUID

/** Persists idempotency keys until a create command succeeds. */
class IdempotencyKeyStore(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    fun keyFor(draftKey: String): String =
        prefs.getString(draftKey, null) ?: UUID.randomUUID().toString().also { fresh ->
            prefs.edit().putString(draftKey, fresh).apply()
        }

    fun clear(draftKey: String) {
        prefs.edit().remove(draftKey).apply()
    }

    companion object {
        private const val PREFS = "moment_create_idempotency"
    }
}

File: momentra/momentra/Data/IdempotencyKeyStore.swift
import Foundation

/// Persists idempotency keys until a create command succeeds.
final class IdempotencyKeyStore {
    private let defaults: UserDefaults
    private let suiteName = "moment_create_idempotency"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func keyFor(draftKey: String) -> String {
        if let existing = defaults.string(forKey: storageKey(draftKey)) {
            return existing
        }
        let fresh = UUID().uuidString
        defaults.set(fresh, forKey: storageKey(draftKey))
        return fresh
    }

    func clear(draftKey: String) {
        defaults.removeObject(forKey: storageKey(draftKey))
    }

    private func storageKey(_ draftKey: String) -> String {
        "\(suiteName).\(draftKey)"
    }
}
